spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java

```diff
@@ -21,22 +21,126 @@
 import java.util.Calendar;
 import java.util.Collection;
 import java.util.Date;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
 import com.google.cloud.ByteArray;
 import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.Type.Code;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
 import com.google.common.collect.ImmutableSet;
 
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerDataException;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerPersistentProperty;
 import org.springframework.core.convert.converter.Converter;
 import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
 
 /**
  * @author Balint Pato
  * @author Chengyuan Zhao
  */
 public class ConversionUtils {
 
+	public static final Map<Type, Class> SPANNER_COLUMN_TYPE_TO_JAVA_TYPE_MAPPING =
+			new ImmutableMap.Builder<Type, Class>()
+			.put(Type.bool(), Boolean.class).put(Type.bytes(), ByteArray.class)
+			.put(Type.date(), com.google.cloud.Date.class)
+			.put(Type.float64(), Double.class).put(Type.int64(), Long.class)
+			.put(Type.string(), String.class)
+			.put(Type.array(Type.float64()), double[].class)
+			.put(Type.array(Type.int64()), long[].class)
+			.put(Type.array(Type.bool()), boolean[].class)
+			.put(Type.timestamp(), Timestamp.class).build();
+
+	private static final Map<Class, Type> JAVA_TYPE_TO_SPANNER_COLUMN_TYPE_MAPPING;
+
+	static {
+		ImmutableMap.Builder<Class, Type> builder = new Builder<>();
+		SPANNER_COLUMN_TYPE_TO_JAVA_TYPE_MAPPING.keySet().stream().forEach(type -> builder
+						.put(SPANNER_COLUMN_TYPE_TO_JAVA_TYPE_MAPPING.get(type), type));
+		JAVA_TYPE_TO_SPANNER_COLUMN_TYPE_MAPPING = builder.build();
+	}
+
+	private static String getTypeDDLString(Type type) {
+		Assert.notNull(type, "A valid Spanner column type is required.");
+		if (type.getCode() == Code.ARRAY) {
+			return "ARRAY<" + getTypeDDLString(type.getArrayElementType()) + ">";
+		}
+		return type.toString()
+				+ (type.getCode() == Code.STRING || type.getCode() == Code.BYTES ? "(MAX)"
```

Comment:
we could add options to specify the size, but using MAX seems ok since it doesn't actually use any extra space in Spanner. however, if the user did want exceptions to be thrown for too-long inputs, then we do lose that functionality. It would be pretty simple to add via an annotation on the persistent property.

Reply:
Let's add the length annotation. It shouldn't be much and then we don't have to touch this again/create a new issue for it, etc.